internal/log/log.go

```diff
@@ -0,0 +1,117 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package log
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"syscall"
+)
+
+var (
+	level            = LevelNone
+	writer  *os.File = os.Stderr
+	writerM sync.Mutex
+
+	context     = make(map[string]string)
+	contextKeys []string
+	contextM    sync.RWMutex
+)
+
+func Close() (err error) {
+	writerM.Lock()
+	defer writerM.Unlock()
+
+	if writer != os.Stderr && writer != os.Stdout {
+		err = writer.Close()
+	}
+	writer = os.Stderr
+
+	contextM.Lock()
+	defer contextM.Unlock()
+	context = make(map[string]string)
+
+	return
+}
+
+func SetLevel(l Level) {
+	level = l
+}
+
+func SetOutput(f *os.File) {
+	writerM.Lock()
+	defer writerM.Unlock()
+
+	writer = f
+}
+
+func SetContext(key string, value string) {
+	contextM.Lock()
+	defer contextM.Unlock()
+
+	if value != "" {
+		if _, found := context[key]; !found {
+			contextKeys = append(contextKeys, key)
+		}
+		context[key] = value
+	} else {
+		delete(context, key)
+		for i := 0; i < len(contextKeys); {
+			if contextKeys[i] == key {
+				contextKeys = append(contextKeys[:i], contextKeys[i+1:]...)
```

Comment:
Not a big deal but you can actually break out early and save some iterations here, since the other branch guarantees that a given context key is only inserted once.